Fix ZRANGESTORE crash when zset_max_listpack_entries is 0

vitarb opened the pull request:
When `zrangestore` is called container destination object is created. 
Before this PR we used to create a listpack based object even if `zset-max-ziplist-entries` or equivalent`zset-max-listpack-entries` was set to 0.
This triggered immediate conversion of the listpack into a skiplist in `zrangestore`, which hits an assertion [here](https://github.com/redis/redis/blob/unstable/src/t_zset.c#L1189) resulting in an engine crash.

Added a TCL test that reproduces this issue.

enjoy-binbin commented on src/t_zset.c:
```diff
@@ -2952,7 +2952,11 @@ static void zrangeResultFinalizeClient(zrange_result_handler *handler,
-    handler->dstobj = createZsetListpackObject();
+    if (server.zset_max_listpack_entries > 0) {
+        handler->dstobj = createZsetListpackObject();
+    } else { // Use skiplist if listpack is disabled to avoid redundant conversion on add.
```
we generally comment like this
```suggestion
    } else {
        /* Use skiplist if listpack is disabled to avoid redundant conversion on add. */
```

vitarb replied:
Done

enjoy-binbin commented on tests/unit/type/zset.tcl:
```diff
@@ -2208,6 +2208,16 @@ start_server {tags {"zset"}} {
         assert_match "*syntax*" $err
     }
 
+    test {ZRANGESTORE skiplist} {
+        set original_max [lindex [r config get zset-max-ziplist-entries] 1]
+        r config set zset-max-ziplist-entries 0
+        r flushall
+        r zadd z1{t} 1 a
+        set res [r zrangestore z2{t} z1{t} 0 -1]
+        assert_equal $res 1
+        r config set zset-max-ziplist-entries $original_max
+    }
+
```
i would like to change like this:
1. let's mention the test purpose in the title
2. avoid using flushall, we can just del z1{t} z2{t}
3. using `zset-max-listpack-entries` the new alias name
```suggestion
    test {ZRANGESTORE with zset-max-listpack-entries 0 dst key should use skiplist encoding} {
        set original_max [lindex [r config get zset-max-listpack-entries] 1]
        r config set zset-max-listpack-entries 0
        r del z1{t} z2{t}
        r zadd z1{t} 1 a
        assert_equal 1 [r zrangestore z2{t} z1{t} 0 -1]
        r config set zset-max-listpack-entries $original_max
    }
```

vitarb replied:
Done

enjoy-binbin reviewed (approved):
verified, thanks for the fix, LGTM

sundb reviewed (commented):
Your changes don't look complete. 
In the module we may still create an empty zset listpack, which is then converted to skiplist when zadd.
Like following code in module
```
    moduleCreateEmptyKey(key,REDISMODULE_KEYTYPE_ZSET);
    zsetAdd(key->value,....); 
```

Perhaps we can allow empty listpack zset to be converted to skiplist zset.

vitarb commented:
Thanks for the feedback, I think I've addressed it:
1. Added similar check into `module.c`
2. In order to prevent crashes added a null check into `zset` allowing empty zset conversion.
3. Addressed styling comments and updated the test.

Technically #2 is sufficient for fixing the issue by itself, but I think we should keep listpack vs ziplist selection logic as an optimization.

sundb commented:
@vitarb Since we already support converting empty listpack zset to skiplist zset, I'm not sure if we still need to determine zset_max_listpack_entries > 0 and create skiplist in two places, I personally don't feel very strongly about it.
And this would cause it to be handled inconsistently with listpack hash and etc.

vitarb commented:
I don't feel strongly about it either, those additional checks aren't needed to prevent the crash anymore given that I've added a NULL check catch-all, however IMO it's nice to have it as a minor optimization to avoid additional imminent conversion into skiplist later.

oranagra commented on src/t_zset.c:
```diff
@@ -1186,23 +1186,23 @@ void zsetConvert(robj *zobj, int encoding) {
-            node = zslInsert(zs->zsl,score,ele);
-            serverAssert(dictAdd(zs->dict,ele,&node->score) == DICT_OK);
-            zzlNext(zl,&eptr,&sptr);
+            while (eptr != NULL) {
+                score = zzlGetScore(sptr);
+                vstr = lpGetValue(eptr, &vlen, &vlong);
+                if (vstr == NULL)
+                    ele = sdsfromlonglong(vlong);
+                else
+                    ele = sdsnewlen((char *) vstr, vlen);
+
+                node = zslInsert(zs->zsl, score, ele);
+                serverAssert(dictAdd(zs->dict, ele, &node->score) == DICT_OK);
+                zzlNext(zl, &eptr, &sptr);
+            }
```
maybe we can easily avoid indenting all these lines by adding an `else` like
```c
        if (eptr != NULL) {
            sptr = lpNext(zl, eptr);
            serverAssertWithInfo(NULL, zobj, sptr != NULL);
        } else {
            eptr = NULL;
        }

or actually, just set it to NULL when it is declared.

vitarb replied:
Assuming you meant `sptr` in the else branch, I don't think we need that assignment at all since it is never accessed if `eptr == NULL`, and if eptr isn't NULL sptr is guaranteed to be not NULL as well.

oranagra commented on src/t_zset.c:
```diff
@@ -2952,7 +2952,12 @@ static void zrangeResultFinalizeClient(zrange_result_handler *handler,
 {
     UNUSED(length);
-    handler->dstobj = createZsetListpackObject();
+    if (server.zset_max_listpack_entries > 0) {
+        handler->dstobj = createZsetListpackObject();
+    } else {
+        /* Use skiplist if listpack is disabled to avoid redundant conversion on add. */
+        handler->dstobj = createZsetObject();
+    }
```
i don't like to repeat this condition everywhere... maybe that signifies that we should have a wrapper method that implements that logic.

but more importantly, i don't like the fact we we fixed zsetConvert (to be on the safe side), but this fix is unreachable / untested, because of these checks on creation.

I see you said you prefer to keep it since it's an optimization, but i think this optimization is unnecessary, it's not a common case / configuration, and the impact is probably low anyway.
converting an empty listpack will be very fast, and the only overhead is the malloc / free.
so i'd rather cleaner code in that case..

p.s. if we do want to optimize the creation, we should be looking at the `length` argument, so when we know the size of the result ahead of time, we can create a skiplist object in advance, instead of converting it when it reaches the threshold (that conversion does have a price)

vitarb replied:
Okay, I'll remove those checks, I don't think that we'll see any noticeable performance impact anyways.
Also moved null check into a separate if statement as you've suggested.

oranagra replied:
FYI: added that other optimization idea here: https://github.com/redis/redis/pull/10789 didn't prove it's any good yet, but didn't wanna forget it..

oranagra reviewed (commented):
Thank you all!
just catching up now, posting my feedback.

vitarb commented:
Btw. I've noticed my IDE (CLion) has different indentation preferences, is there any shared codestyle that I can use?

sundb commented:
> Btw. I've noticed my IDE (CLion) has different indentation preferences, is there any shared codestyle that I can use?

Use 4-space indentation.
There is no fixed codestyle and the best way to do this is to refer to older code.

vitarb commented:
He folks, unless there is more feedback, can someone with write access approve this?
